File: january2025/january12.py

class Solution:
    def maxWater(self, arr):
        # code here
        if not arr:
            return 0
        n=len(arr)
        leftmax=[0]*n
        rightmax=[0]*n
        
        #left maximum for the array
        leftmax[0]=arr[0]
        for i in range(1,n):
            leftmax[i]=max(leftmax[i-1],arr[i])
            
        #right max
        rightmax[n-1]=arr[n-1]
        for i in range(n-2,-1,-1):
            rightmax[i]=max(rightmax[i+1],arr[i])
        
        # finding the water trapped
        water=0
        for i in range(n):
            water+=min(leftmax[i],rightmax[i])-arr[i]
            
        
        return water

#{ 
 # Driver Code Starts
#Initial template for Python 3

import math


def main():
    t = int(input())
    while (t > 0):

        arr = [int(x) for x in input().strip().split()]
        obj = Solution()
        print(obj.maxWater(arr))

        t -= 1
        print("~")


if __name__ == "__main__":
    main()

# } Driver Code Ends

File: january2025/january12.java
class Solution {
    public int maxWater(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }

        int n = height.length;
        int[] leftMax = new int[n];
        int[] rightMax = new int[n];

        // Calculate left maximum for each index
        leftMax[0] = height[0];
        for (int i = 1; i < n; i++) {
            leftMax[i] = Math.max(leftMax[i - 1], height[i]);
        }

        // Calculate right maximum for each index
        rightMax[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            rightMax[i] = Math.max(rightMax[i + 1], height[i]);
        }

        // Calculate water trapped at each index
        int water = 0;
        for (int i = 0; i < n; i++) {
            water += Math.min(leftMax[i], rightMax[i]) - height[i];
        }

        return water;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();

        while (t > 0) {
            int n = scanner.nextInt();
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = scanner.nextInt();
            }

            Solution obj = new Solution();
            System.out.println(obj.maxWater(arr));

            t--;
        }
    }
}
